packages/patterns/src/signal.ts

```diff
@@ -31,41 +32,61 @@ export type Listener<T> = (data: T) => void;
  * Notice that the Signals are public.
  * We don't need to implement specific subscriptions on the class, unless we need to expose it as a remote service.
  */
-export class Signal<T> extends Set<Listener<T>> {
-    private onceHandlers = new Set<Listener<T>>();
+export class Signal<T> {
+    private handlers = new Map<Listener<T>, IS_ONCE>();
+    constructor(handlers?: Listener<T>[], once?: Listener<T>[]) {
+        handlers?.forEach((handler) => this.subscribe(handler));
+        once?.forEach((handler) => this.once(handler));
+    }
+
     /**
      * Subscribe a notification callback
      * @param handler - Will be executed with a data arg when a notification occurs
      */
     subscribe = (handler: Listener<T>) => {
-        this.add(handler);
+        this.handlers.set(handler, false);
     };
+
+    /**
+     * Subscribe to only the next notification
+     * @param handler - Will be executed with a data arg when a notification occurs
+     */
     once = (handler: Listener<T>) => {
-        this.onceHandlers.add(handler);
+        this.unsubscribe(handler);
```

Comment:
Why is the unsubscribe needed?

Reply:
to ensure the order is affected when re-adding an existing handler.
There are a few behaviours that can happen:
1. ignore duplication (may be unclear when changing a persistent handler to "once" handler)
2. add duplicates (nope)
3. change order only when changing a persistent handler to "once" handler (IMO, confusing)
4. always make the latest added handler be called last, even if it's a duplicate (chosen)

Reply:
I'm not sure, but I think `#1` would be my pick. If I call `addListener` multiple times it usually has no effect, I think that modifying a handler to only run once should either just change the flag without changing the order.

Alternatively it would be even better if it would throw if I try to change from a repeating handler to a one-time handler (not the other way around) and allow to register a one-time handler with force to explicitly bypass the error.

Reply:
changed as per your suggestion